### PR TITLE
mya: init at 1.0.0

### PR DESCRIPTION
**Pull request by pbsds**
###### Description of changes

Add `mya`, [myanimelist-cli](https://github.com/jmakhack/myanimelist-cli/).
The only headache when using it is https://github.com/jmakhack/myanimelist-cli/issues/112, since they don't provide a default api key i ~~added a way to set it using an override~~, document the environment variable to set in `meta.longDescription`.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/misc/myanimelist-cli/default.nix
Comment (nevivurn) [inline pkgs/applications/misc/myanimelist-cli/default.nix:34]: We should omit this. I'm not 100% sure how sensitive the client ID is, but the environment variable seems perfectly reasonable, no need to make the client ID world-readable in the nix store.

**Review by nevivurn — CHANGES_REQUESTED** [review]
Result of `nixpkgs-review pr 240405` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>myanimelist-cli</li>
  </ul>
</details>

File: pkgs/applications/misc/myanimelist-cli/default.nix
Comment (pbsds) [reply pkgs/applications/misc/myanimelist-cli/default.nix:34]: I removed it, but moved the and rephrased the comment documenting the env variable to lie near the meta block.

**Review by nevivurn — CHANGES_REQUESTED** [review]
Missed this the first time around, build seems to fail on Darwin due to the Makefile looking for `gcc` specifically. Though I can't test darwin, this might work:

```nix
  postPatch = ''
    substituteInPlace Makefile \
      --replace 'gcc' '${stdenv.cc.targetPrefix}cc'
  '';
```

If it can't be made to build, we should set `platforms = platforms.linux` so CI doesn't try to build it.

**Comment by pbsds** [comment]
Now `pkgsLLVM.myanimelist-cli` should build on nixos

@ofborg build myanimelist-cli

**Review by nevivurn — APPROVED** [review]

File: pkgs/applications/misc/myanimelist-cli/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/misc/myanimelist-cli/default.nix:41]: Please move this to longDescription

Comment (drupol) [reply pkgs/applications/misc/myanimelist-cli/default.nix:41]: @pbsds Do you plan to implement this feedback?

File: pkgs/applications/misc/myanimelist-cli/default.nix
Comment (drupol) [inline pkgs/applications/misc/myanimelist-cli/default.nix:9]: I think we should use the `finalAttrs` pattern here.

The `finalAttrs` pattern will let you remove the `rec` keyword ([see its implementation in Nix](https://github.com/NixOS/nix/blob/2d1d81114d72ace89ce08cd3bc93f4eb27a2975d/src/libexpr/nixexpr.cc#L121)).

Why is this pattern preferred to `rec` ?

Let's take this simple code example:

```nix
mkDerivation rec {
   foo = 1;
   bar = foo + 1;
}
```

and then `.overrideAttrs(old: { foo = 2; })`, you'll get `{ foo = 2; bar = 2; }` while with `finalAttrs` pattern, it would work correctly because it's a real fixed point.

Let me share a couple of useful links regarding the `finalAttrs` pattern:

1. History: https://discourse.nixos.org/t/avoid-rec-expresions-in-nixpkgs/8293
2. Documentation: https://nixos.org/manual/nixpkgs/unstable/#mkderivation-recursive-attributes
3. Recent example of implementation: https://github.com/NixOS/nixpkgs/compare/17f96f7b978e61576cfe16136eb418f74fab9952..9e6ea843e473d34d4f379b8b0d8ef0425a06defe

Feel free to reach out if you need some assistance.

**Review by drupol — CHANGES_REQUESTED** [review]

File: pkgs/applications/misc/myanimelist-cli/default.nix
Comment (drupol) [inline pkgs/applications/misc/myanimelist-cli/default.nix:55]: Dear @NixOS/darwin-maintainers, can you help on this?

**Review by drupol — COMMENTED** [review]
Thanks for implementing the requested changes.

It would be nice to have it working on Darwin as well, I'm pretty sure this is something we can do here.

File: pkgs/applications/misc/myanimelist-cli/default.nix
Comment (drupol) [reply pkgs/applications/misc/myanimelist-cli/default.nix:55]: @pbsds Have a look at this: https://github.com/NixOS/nixpkgs/pull/194312 it might probably be the solution.

Comment (emilazy) [reply pkgs/applications/misc/myanimelist-cli/default.nix:55]: Yes, an `argp-standalone` build input is the way to go. Ideally it should be conditioned on "libc is not glibc" rather than Darwin specifcially (for instance, it applies to other BSDs), but I'm not sure if there's a convenient way to express this. If that isn't enough I can take another look.

Comment (emilazy) [reply pkgs/applications/misc/myanimelist-cli/default.nix:55]: Maybe `lib.optionals (!stdenv.hostPlatform.isGnu)` is the right logic.

**Comment by pbsds** [comment]
Marking this as draft while i abuse ofborg

**Comment by emilazy** [comment]
Wow, it uses both `libbsd` and argp...

I will try to get this working and post a patch today.

**Comment by pbsds** [comment]
There seems to possibly be [an upstream story](https://gitlab.freedesktop.org/libbsd/libbsd/-/merge_requests/3). The darwin patches to libbsd seem to only disable system-provided functions.

**Comment by emilazy** [comment]
That was fun. So the upstream use of `libbsd` here isn't ideal (forgoing overlay mode in favour of explicit includes), and our `libbsd` package is also a bit broken on macOS right now. I think it would have been possible to just drop the dependency on `libbsd` for Darwin, but I went to the effort of fixing `libbsd` instead to avoid touching the source more than necessary. I'll open a PR for the `libbsd` changes, but here's the diff against the current PR head that it took to get this working; the changes should be harmless on Linux/glibc.

Additionally, I switched to using the CMake build system as it seems that upstream is using it for CI. Unfortunately it doesn't link `argp-standalone` like the Makefile does, so I had to add a patch for that. It should probably go upstream (and I hereby release it into the public domain and/or license it under [0BSD](https://opensource.org/license/0bsd/) to make that zero-friction in case I don't get around to it and you feel like doing so). I also changed the `argp-standalone` conditional to be a bit more precise for any Nix-on-musl-or-BSD weeaboos out there.

(I wonder if we should call this `mya` instead, to match the binary and AUR package name?)

```diff
diff --git a/pkgs/applications/misc/myanimelist-cli/argp.patch b/pkgs/applications/misc/myanimelist-cli/argp.patch
new file mode 100644
index 000000000000..352f8a077adc
--- /dev/null
+++ b/pkgs/applications/misc/myanimelist-cli/argp.patch
@@ -0,0 +1,26 @@
+diff --git a/CMakeLists.txt b/CMakeLists.txt
+index 273968c..236e5fb 100644
+--- a/CMakeLists.txt
++++ b/CMakeLists.txt
+@@ -3,10 +3,6 @@ project(mya)
+ set(CMAKE_C_STANDARD 11 )
+ set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -g -Wall")
+ 
+-if(UNIX)
+-	set(LINUX TRUE)
+-endif()
+-
+ set(SRC_DIR src)
+ set(INC_DIR include)
+ 
+@@ -17,7 +13,8 @@ set_target_properties(mya PROPERTIES RUNTIME_OUTPUT_DIRECTORY "${CMAKE_BINARY_DI
+ target_include_directories(mya PRIVATE ${INC_DIR})
+ 
+ set(LIBS curl json-c bsd)
+-if(LINUX)
+-    list(APPEND LIBS)
++find_library(ARGP argp)
++if(ARGP)
++    list(APPEND LIBS argp)
+ endif()
+ target_link_libraries(mya ${LIBS})
diff --git a/pkgs/applications/misc/myanimelist-cli/default.nix b/pkgs/applications/misc/myanimelist-cli/default.nix
index e4e15ff66b06..10aaaf39dda8 100644
--- a/pkgs/applications/misc/myanimelist-cli/default.nix
+++ b/pkgs/applications/misc/myanimelist-cli/default.nix
@@ -1,10 +1,12 @@
 { lib
 , stdenv
 , fetchFromGitHub
-, argp-standalone
+, cmake
+, ninja
 , curl
 , json_c
 , libbsd
+, argp-standalone
 }:
 
 stdenv.mkDerivation (finalAttrs: {
@@ -18,24 +20,31 @@ stdenv.mkDerivation (finalAttrs: {
     hash = "sha256-EmdkPpYEUIk9hr6rbnixjvznKSEnTCSMZz/17BfHGCk=";
   };
 
+  nativeBuildInputs = [
+    cmake
+    ninja
+  ];
+
   buildInputs = [
     curl
     json_c
     libbsd
-  ] ++ lib.optionals stdenv.isDarwin [
+  ] ++ lib.optionals (!stdenv.hostPlatform.isGnu) [
     argp-standalone
   ];
 
-  # support LLVM and such
-  postPatch = ''
-    substituteInPlace Makefile \
-      --replace 'CC:=gcc' 'CC:=${stdenv.cc.targetPrefix}cc'
-  '';
+  patches = [
+    ./argp.patch
+  ];
 
   installPhase = ''
     runHook preInstall
-    mkdir -p $out
+
+    # Based on the upstream PKGBUILD
+    mkdir -p $out/share/doc/${finalAttrs.pname}
     cp -a bin $out
+    cp $cmakeDir/README.md $out/share/doc/${finalAttrs.pname}
+
     runHook postInstall
   '';
 
@@ -54,7 +63,5 @@ stdenv.mkDerivation (finalAttrs: {
     maintainers = with maintainers; [ pbsds ];
     mainProgram = "mya";
     platforms = platforms.all;
-    # FIXME: darwin maintainer needed
-    #broken = stdenv.isDarwin; # src/mya.c:4:10: fatal error: 'argp.h' file not found
   };
 })
```

**Comment by pbsds** [comment]
Amazing! I pushed your patch, co-authoring you.

> (I wonder if we should call this mya instead, to match the binary and AUR package name?)

That was my original pname, but then I assumed the repo name was canon, never checking repology. I switched the pname to `mya`.

@ofborg build mya

**Comment by emilazy** [comment]
Yay! The Darwin build will still expectedly fail until I get these `libbsd` changes merged. By the way, it looks like you didn't adjust the `pname`.

**Review by emilazy — APPROVED** [review]

**Pull request merged**
